### ISSUE TITLE
Replacing graph data loses track of nodes

### ISSUE THREAD

**Issue by mgfcf** [vasturiano/force-graph#216]
Hey there! Really great graph visualization tool! I am working on a graph editor that uses the visualization as an interface. Implementing an undo/redo feature seems to be a bit harder, since when I try to update the data with `.graphData(..)` I get the following error:

```
Uncaught Error: node not found: undefined
    Oi https://unpkg.com/force-graph:5
    d https://unpkg.com/force-graph:5
    links https://unpkg.com/force-graph:5
    update https://unpkg.com/force-graph:5
    s https://unpkg.com/force-graph:2
    u https://unpkg.com/force-graph:2
    setTimeout handler*l https://unpkg.com/force-graph:2
    Yr/</</i[t.name]</< https://unpkg.com/force-graph:2
    onChange https://unpkg.com/force-graph:5
    Yr/</</i[t.name]</< https://unpkg.com/force-graph:2
    *My call to .graphData(..)*
```

**To Reproduce**
How I manipulate the graph:
1. Initial state is loaded
 - State gets saved
2. I add a link
 - Here a new state gets saved
3. I undo
 - Here, the original graph seems be restored successfully, even tho the node colors have been lost
4. I redo
5. I undo
 - Now all the links are gone and I get exceptions

How a link is added:
1. A new link is created: link = {source: "id", target: "id"}
 - Pointing out: I only use ids, not full node objects in source and target. The graph seems handle that fine and updates it internally.
2. Link is added to existing graph data
3. Graph data is reapplied: `graphObj.graphData(data);`
 - This always works fine and never created any issues

How a state is saved:
1. I create a shallow copy of the graph data in the form of: `{ links: [ {source: "id", target: "id", "index": 0}, ...], nodes: [{ id: "id", index: 0, x: 0, y: 0, vx: 0, vy: 0, fx: 0, fy: 0}, ...]}`. I make sure to copy all the simulation data, since I don't want the graph to simulate anymore and stay in it's configuration.
2. I convert this object to a JSON string and store the string in the history. I do this, to create a deep copy, so that future changes to the graph don't affect the stored states.

How a state is undone:
1. I parse the previous JSON data and get a brand new data object.
2. I set the new data with `graphObj.graphData(data);`.

**Expected behavior**
The graph changes to a previous state.

**Screenshots**
Initial state when saved:
![image](https://user-images.githubusercontent.com/24680560/127528462-5e91ed5a-e510-4b56-9a40-3a59f9101901.png)

Initial state when reloaded:
![image](https://user-images.githubusercontent.com/24680560/127528756-7974f28d-9e2a-4370-841b-c1ff925fa529.png)

**Additional context**
Same happens if I go back to the very first state, right when the animation has been loaded, before physics kicked in.

Seems like the graph loses track of the nodes and doesn't know what to do with the links, would be why they aren't rendered.

Please let me know, how I can help out with more information. I could also supply JSON state objects.

Thanks for all the help in advance,
Max

**Comment by vasturiano** [comment]
Thanks for reaching out Max.

Normally this error occurs if at least one of your links reference a node (in `source` or `target`) which doesn't exist in the `nodes` structure. In fact, normally this arises from the underlying graph library [d3-force](https://github.com/d3/d3-force).

To try to understand and debug a bit better what's happening what would be helpful is if you could make a simple example that reproduces the error, on [codepen](https://codepen.io/) or similar.

**Comment by mgfcf** [comment]
Hey! Thanks for your reply.

Took me some time, but I finally got around to reproduce the error in a codepen:
[https://codepen.io/MaxUser/pen/yLXaoGm](https://codepen.io/MaxUser/pen/yLXaoGm)

I hope the code is documented enough, but feel free to ask for clarifications. I used relevant code snippets from the project that I am working on recreated a basic editor. Tho so far the error only occurs when you try to use the "connect nodes" feature. In the original project, the error seems to occur more frequently, but I was unable to reproduce it on a smaller scale so far.

My project also saves the very initial state of the graph, which also causes issues when you try to undo back to it. I left that out tho, since that seems like an edge case and can easily be avoided.

For some reason, the important errors don't show up in the codepen internal console for me, so I recommend using the browser built-in one.

And yes, as you mentioned, the errors seem to originate in the d3-force library.

I tried to print out all the relevant information in the console. The history is stored in an array, with the first element being the newest save, and higher indices being older saves. Every undo or redo only loads the stored json text from the history into a json object and loads it into the graph. The objects stored in the history seem fine and work as valid graphs, when I supply this data as initial graph date. So I assume that something breaks in the library overtime, with all the changes happening, which leads to newer graph data not working anymore. Also notice, how nothing works anymore, after the error occurred once. No new data will work.

Thank you very much, for helping me understand this issue!
Max

**Comment by vasturiano** [comment]
@mgfcf thanks for making the example!

I've tried it and I seem to get an error already when clicking `undo` the first time. I think it's possibly related to how you're serializing and unserializing the data. The error is on this line:
```
cleanLink.source = link.source.id;
```

This happens right after the data is unserialized, but at this point in time the force engine still didn't have the chance to run at least for one iteration, which is when the `source` and `target` attributes in the links are converted from pure node ids into node object references. Thus, at this moment `link.source` is still a string (node id), which leads into the error: `Cannot read property 'id' of undefined`.

I'm not sure if there are more issues with the process, this is the one I run into first. Hope this helps somewhat!

**Comment by mgfcf** [comment]
@vasturiano thank you for taking a look at it.

The error that you mentioned is my mistake, since the mentioned line only works when the graph has been successfully initialized. Usually the line will be executed _after_ that happened, tho apparently the order somehow changes when the graph cannot be updated in the force-graph library.

The mentioned code was only added to provide some output to the console. I removed the unnecessary code, so that only the problem in the library should be happening.

Could you please take another look at it? I would be very thankful for that.
[https://codepen.io/MaxUser/pen/yLXaoGm](https://codepen.io/MaxUser/pen/yLXaoGm)

To make sure "normal" behavior works as expected, you can try to add and remove some nodes at first and try to undo and redo you edits.

Thank you very much for your time.
Max

**Comment by vasturiano** [comment]
@mgfcf thanks for updating the code.

I'm still seeing an issue occurring right the first time that "undo" is pressed.

The error occurs because you are including an empty link (last item in the links array here), which misses the mandatory attributes `source` and `target`:
<img width="578" alt="Screen Shot 2021-09-10 at 12 27 40 PM" src="https://user-images.githubusercontent.com/6784226/132850702-37501400-0426-4075-a451-07f17fdc96a8.png">

This empty item got generated when serializing the data at each of the steps:
<img width="383" alt="Screen Shot 2021-09-10 at 12 58 16 PM" src="https://user-images.githubusercontent.com/6784226/132850819-7a2c71ba-17d5-4d99-8abb-1f5fdb3f4cd4.png">

Mainly because of the same line as before, where you expect to find an object in `link.source`/`link.target`, but it's still a string:
<img width="546" alt="Screen Shot 2021-09-10 at 1 00 45 PM" src="https://user-images.githubusercontent.com/6784226/132850959-32cf5453-f415-472c-98b8-37df8c521538.png">

**Comment by mgfcf** [comment]
Oh wow! Thank you so much. I am sorry that I didn't notice that myself @vasturiano !

I was trying to convert the graph to the so called "clean" version, right after I updated it on the graph. Hoping, that it would have taken the simple json data (links in the format {source: NODE_ID, target: NODE_ID}) and converted it to the more complex working version (links in the format {source: {id: NODE_ID}, target: {id: NODE_ID}}), where source and target are not just the node id, but the node object. At least in the case of links, this does not happen immediately tho. Adding a short timeout of 5 seconds, after the graph data has been loaded and before the "clean" version get saved shows, that the graph is actually not converting it immediately, but (probably) waiting for some idle time to do it. So every time I tried to access the more complex objects directly after loading it into the graph (when adding a link) the data would somehow not be converted immediately, at least not for the newly added link.

I am certainly not familiar with the inner workings of the force-graph library, but something funky is apparently going on with the execution order. A little different then I expected.

Since I now know how to fix my issue, I will close this thread. Thank you very much @vasturiano for your support, and for providing such active help in this repository.

Max
